js/src/components/paid-ads/billing-card/billing-setup-card.test.js

```diff
@@ -0,0 +1,114 @@
+import BillingSetupCard from './billing-setup-card';
+import useWindowFocus from '.~/hooks/useWindowFocus';
+
+jest.mock( '.~/hooks/useGoogleAdsAccount', () =>
+	jest.fn().mockName( 'useGoogleAdsAccount' ).mockReturnValue( {} )
+);
+
+jest.mock( '.~/hooks/useDispatchCoreNotices', () =>
+	jest
+		.fn()
+		.mockName( 'useDispatchCoreNotices' )
+		.mockReturnValue( { createNotice: () => {} } )
+);
+
+jest.mock( '.~/hooks/useWindowFocus', () =>
+	jest.fn().mockName( 'useWindowFocus' ).mockReturnValue( true )
+);
+
+jest.mock( '@wordpress/api-fetch', () => {
+	const impl = jest.fn().mockName( '@wordpress/api-fetch' );
+	impl.use = jest.fn().mockName( 'apiFetch.use' );
+	return impl;
+} );
+
+describe( 'BillingSetupCard', () => {
+	let fetchBillingStatus;
+
+	beforeEach( () => {
+		fetchBillingStatus = jest
+			.fn()
+			.mockResolvedValue( { status: 'approved' } );
+
+		apiFetch.mockImplementation( ( { path } ) => {
+			switch ( path ) {
+				case '/wc/gla/ads/accounts':
+					return Promise.resolve( {} );
+
+				case '/wc/gla/ads/billing-status':
+					return fetchBillingStatus();
+
+				default:
+					throw new Error( `No mocked apiFetch for path: ${ path }` );
+			}
+		} );
+	} );
+
+	it( 'Should only call back `onSetupComplete` once regardless of its reference change', async () => {
+		const inspect = jest.fn();
+
+		let onSetupComplete = () => inspect();
+		let rerender;
+
+		// Initial with a not yet set up billing.
+		fetchBillingStatus.mockResolvedValueOnce( { status: 'unknown' } );
+		await act( async () => {
+			const result = render(
+				<BillingSetupCard onSetupComplete={ onSetupComplete } />
+			);
+			rerender = result.rerender;
+		} );
+
+		expect( fetchBillingStatus ).toHaveBeenCalledTimes( 1 );
+		expect( inspect ).toHaveBeenCalledTimes( 0 );
+
+		// Simulate document regains focus to trigger billing status check again.
+		useWindowFocus.mockReturnValueOnce( false );
+		rerender( <BillingSetupCard onSetupComplete={ onSetupComplete } /> );
+		await act( async () => {
+			rerender(
+				<BillingSetupCard onSetupComplete={ onSetupComplete } />
+			);
+		} );
+
+		expect( fetchBillingStatus ).toHaveBeenCalledTimes( 2 );
+		expect( inspect ).toHaveBeenCalledTimes( 1 );
+
+		// Change the reference of callback.
+		onSetupComplete = () => inspect();
+		await act( async () => {
+			rerender(
+				<BillingSetupCard onSetupComplete={ onSetupComplete } />
+			);
+		} );
+
+		expect( inspect ).toHaveBeenCalledTimes( 1 );
+	} );
+
+	it( 'Should not call back `onSetupComplete` again by the next interval after billing is already approved', async () => {
+		const onSetupComplete = jest.fn();
+
+		await act( async () => {
+			render( <BillingSetupCard onSetupComplete={ onSetupComplete } /> );
+		} );
+
+		expect( fetchBillingStatus ).toHaveBeenCalledTimes( 1 );
+		expect( onSetupComplete ).toHaveBeenCalledTimes( 1 );
+
+		await act( async () => {
+			jest.runOnlyPendingTimers();
+		} );
+
+		expect( fetchBillingStatus ).toHaveBeenCalledTimes( 2 );
```

Comment:
I would assume that once the billing status is approved, there's no need to call fetchBillingStatus again since it becomes unnecessary. WDYT?

Reply:
The `fetchBillingStatus` being called again is the internal implementation logic of `useAutoCheckBillingStatusEffect` hook, and yes, it's not needed for meeting the same business logic. However, the hook naming does not only scope the need for `checkStatusAndCompleteSetup` but also any possible changes of billing status, although there has been no that kind of need from the beginning till now.

From a test case perspective, it's not required to assert it. Removed by https://github.com/woocommerce/google-listings-and-ads/pull/2127/commits/f8e81083cee0b2c8dfacb0dea2652308cfe7a6f9.